[cond] make cond not throw warnings on constant pred in eager mode

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #138837
 We don't raise warnings for torch.cond in eager mode the motivation is in  https://github.com/pytorch/pytorch/issues/138782. 


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

ydwu4 committed:
[cond] make cond not throw warnings on constant pred in eager mode

[ghstack-poisoned]

zou3519 commented on torch/_higher_order_ops/cond.py:
```diff
@@ -137,10 +138,12 @@ def false_fn(x: torch.Tensor):
-            "Pred is a Python constant. When used with torch.cond, it executes only one of the branches."
-            " If you want torch.cond to perserve two branches, please make the predicate a boolean tensor or a SymBool."
-        )
+        if torch.compiler.is_compiling():
```
Can you add a comment here saying this is the non-strict export case?

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -680,9 +681,10 @@ def call_function(
-            log.warning(
-                "Pred is a Python constant. When used with torch.cond, it executes only one of the branches."
-                " If you want torch.cond to perserve two branches, please make the predicate a boolean tensor or a SymBool."
+            warnings.warn(
+                "Pred is a Python constant. When exporting torch.cond, it exports one of the branches."
+                " If you want torch.cond to perserve two branches, please make the predicate a boolean tensor or a SymBool.",
+                UserWarning,
```
The previous error message was better? Since this is the compile path, not the export path.

We should say that we're "specializing" on one of the branches. Also "perserve -> preserve".

zou3519 commented on torch/_higher_order_ops/cond.py:
```diff
@@ -137,10 +138,12 @@ def false_fn(x: torch.Tensor):
-        )
+        if torch.compiler.is_compiling():
+            warnings.warn(
+                "Pred is a Python constant. When exporting torch.cond, it exports one of the branches."
```
ditto here: let's make it clearer that we're just going to specialize on one of the branches.

ydwu4 committed:
Update on "[cond] make cond not throw warnings on constant pred in eager mode"

 We don't raise warnings for torch.cond in eager mode the motivation is in  https://github.com/pytorch/pytorch/issues/138782. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

ydwu4 committed:
Update on "[cond] make cond not throw warnings on constant pred in eager mode"

 We don't raise warnings for torch.cond in eager mode the motivation is in  https://github.com/pytorch/pytorch/issues/138782. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge

ydwu4 committed:
Update on "[cond] make cond not throw warnings on constant pred in eager mode"

 We don't raise warnings for torch.cond in eager mode the motivation is in  https://github.com/pytorch/pytorch/issues/138782. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames rec

[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge
